nav2_controller/src/controller_server.cpp

```diff
@@ -458,10 +458,17 @@ void ControllerServer::computeControl()
       }
 
       if (action_server_->is_cancel_requested()) {
-        RCLCPP_INFO(get_logger(), "Goal was canceled. Stopping the robot.");
-        action_server_->terminate_all();
-        publishZeroVelocity();
-        return;
+        if (controllers_[current_controller_]->cancel()) {
+          RCLCPP_INFO(get_logger(), "Cancellation was successful. Stopping the robot.");
+          action_server_->terminate_all();
+          publishZeroVelocity();
+          return;
+        } else {
+          RCLCPP_INFO_THROTTLE(
+            get_logger(),
+            *get_clock(), 1000,
+            "Goal was canceled. Requesting controller to stop robot gracefully.");
```

Comment:
Nitpick: this is an implementation detail. "Waiting for the controller to finish cancellation"

Reply:
On the other hand, `publishZeroVelocity();` enforces a stop, so this is the only option :smile:. You can leave it.

Reply:
Sure, @Rayman can you make that string update? Then I can merge

Reply:
Done